onnxruntime/python/onnxruntime_pybind_ortvalue.cc

```diff
@@ -0,0 +1,216 @@
+        ORT_ENFORCE(ml_value->IsTensor(), "Only OrtValues that are Tensors are currently supported");
+
+        auto* tensor = ml_value->GetMutable<Tensor>();
+
+        if (tensor->Shape().Size() == 0) {
+          return 0;
+        }
+
+        // Should cover x86 and x64 platforms
+        return reinterpret_cast<int64_t>(tensor->MutableDataRaw());
+      })
+      .def("device_name", [](const OrtValue* ort_value) -> std::string {
+        if (ort_value->IsTensor()) {
+          return std::string(GetDeviceName(ort_value->Get<Tensor>().Location().device));
+        } else {
+          ORT_THROW("Only OrtValues that are Tensors are currently supported");
+        }
+      })
+      .def("shape", [](const OrtValue* ort_value) -> py::list {
+        // OrtValue can only be a Tensor/SparseTensor, make this generic to handle non-Tensors
+        ORT_ENFORCE(ort_value->IsTensor() || ort_value->IsSparseTensor(),
+                    "Only OrtValues that are Tensors/SpareTensors are currently supported");
+
+        py::list shape_arr;
+        const auto& dims = (ort_value->IsTensor())
+                               ? ort_value->Get<Tensor>().Shape().GetDims()
+                               : ort_value->Get<SparseTensor>().Shape().GetDims();
+
+        for (auto dim : dims) {
+          // For sequence tensors - we would append a list of dims to the outermost list
+          // For now only tensors are supported in OrtValue
+          shape_arr.append(dim);
+        }
+
+        return shape_arr;
+      })
+      .def("data_type", [](const OrtValue* ort_value) -> std::string {
+        const ONNX_NAMESPACE::TypeProto* type_proto;
+        // Handle gutless types first to get the actual type
+        if (ort_value->IsTensor()) {
+          auto elem_type = ort_value->Get<Tensor>().GetElementType();
+          type_proto = DataTypeImpl::TensorTypeFromONNXEnum(elem_type)->GetTypeProto();
+        } else if (ort_value->IsSparseTensor()) {
+          auto elem_type = ort_value->Get<SparseTensor>().Values().GetElementType();
+          type_proto = DataTypeImpl::SparseTensorTypeFromONNXEnum(elem_type)->GetTypeProto();
+        } else if (ort_value->IsTensorSequence()) {
+          auto elem_type = ort_value->Get<TensorSeq>().DataType()->AsPrimitiveDataType()->GetDataType();
+          type_proto = DataTypeImpl::SequenceTensorTypeFromONNXEnum(elem_type)->GetTypeProto();
+        } else {
+          // Plane sequences and maps probably have their specific type
+          type_proto = ort_value->Type()->GetTypeProto();
+        }
+
+        ORT_ENFORCE(type_proto != nullptr, "Unknown type of OrtValue: ", ort_value->Type());
+
+        return *ONNX_NAMESPACE::Utils::DataTypeUtils::ToType(*type_proto);
+      })
+      .def("is_tensor", [](const OrtValue* ort_value) -> bool {
+        return ort_value->IsTensor();
+      })
+      .def("is_sparse_tensor", [](const OrtValue* ort_value) -> bool {
```

Comment:
sorry if I am missing something very obvious - but how does a user create an OrtValue in Python that is a sparse tensor or a sequence tensor ? Looking at it, it looks like they can only get hold of such OrtValues, fromwhat is returned from IoBinding ?

Reply:
They will be returned from run call, although this is mostly coming in the next iteration, when OrtValue will contain SparseTensor. TensorSeq is actually supported now but apparently not via python